### PR TITLE
feat: yggdrasil dotnet custom strategies

### PR DESCRIPTION
**Pull request by daveleek**
## About the changes

Implements custom strategy handling in Yggdrasil .NET.
Adds some tests
Improves build of Yggdrasil.NET Engine

### CODE REVIEW

File: dotnet-engine/Yggdrasil.Engine/Types.cs
Comment (daveleek) [reply dotnet-engine/Yggdrasil.Engine/Types.cs:110]: That's a very good point, I'll see if I can provoke the behavior a bit with some unit tests!

File: dotnet-engine/Yggdrasil.Engine.Tests/Utils.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine.Tests/Utils.cs:12]: Nit: Can we move this to the top so we can have methods together and fields together?

File: dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs:18]: What's the plan here?

If we want to keep this, let's leave it in. If it needs refactor to make this easier and saner to use, let's do that in another PR

```suggestion
    [Test]

```

File: dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs:106]: I don't think this is a fail, this should degrade gracefully

```suggestion
    public void Custom_Strategies_Required_But_Not_Configured_Returns_False() {

```

File: dotnet-engine/Yggdrasil.Engine/Types.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine/Types.cs:30]: This looks a little suspect to my eyes, this should never be null, right?

File: dotnet-engine/Yggdrasil.Engine/Types.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine/Types.cs:51]: Cool! That's much better than what I wanted to do!

File: dotnet-engine/Yggdrasil.Engine/Types.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine/Types.cs:70]: Also doesn't look right to me, this should be set if this object exists

File: dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs
Comment (sighphyre) [inline dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs:30]: Okay, I see this is a pattern we're using elsewhere but this needs to change in another PR. Nil pointers are a serious problem since the underlying FFI layer can't return those. So something very, very wrong has happened and we need to know about it

I want to say we should draw inspiration from the Java engine here and have a single method can unwrap the incoming FFI JSON bodies and deal with them

**Review by sighphyre — CHANGES_REQUESTED** [review]
The underlying algorithm looks correct to me but there's a bunch of new warnings and some suspect nullable types that I would like to resolve before we merge this

File: dotnet-engine/Yggdrasil.Engine.Tests/custom-strategies.json
Comment (gastonfournier) [reply dotnet-engine/Yggdrasil.Engine.Tests/custom-strategies.json:1]: Yes, I already did the custom-strategies-test inside Java, I could benefit from having this one merged

File: dotnet-engine/Yggdrasil.Engine.Tests/Utils.cs
Comment (daveleek) [reply dotnet-engine/Yggdrasil.Engine.Tests/Utils.cs]: Done!

File: dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs
Comment (daveleek) [reply dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs:18]: The plan is to ensure that our builds run quickly while we look at a good way to separate tests to be run at build and tests to be run when explicitly invoked, and that seems to be test categories but I wanted to save that for a new PR when I've had a chance to try it out properly

File: dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs
Comment (daveleek) [reply dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs:30]: Yeah I agree, and yeah lets open a new PR with that fix

File: dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs
Comment (gastonfournier) [inline dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs:60]: I think you should move this after taking the state and error handling.

Comment (daveleek) [reply dotnet-engine/Yggdrasil.Engine/YggdrasilEngine.cs:60]: Good point!

File: dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs
Comment (sighphyre) [reply dotnet-engine/Yggdrasil.Engine.Tests/YggdrasilEngineTest.cs:18]: Can we remove the commented out stuff then?

Not gonna block this for that though

**Review by sighphyre — APPROVED** [review]
G

**Pull request merged**
